internal/cf/resourcecache.go

```diff
@@ -0,0 +1,386 @@
+	}
+}
+
+// -----------------------------------------------------------------------------------------------
+// Bindings
+// -----------------------------------------------------------------------------------------------
+
+// addBindingInCache stores a binding to the cache
+func (c *resourceCache) addBindingInCache(key string, binding *facade.Binding) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	c.bindings[key] = binding
+	// TODO :remove After internal review
+	fmt.Printf("Added the binding to Cache: %v \n", binding)
+}
+
+// deleteBindingFromCache deletes a specific binding from the cache
+func (c *resourceCache) deleteBindingFromCache(key string) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	delete(c.bindings, key)
+	// TODO :remove After internal review
+	fmt.Printf("Added the binding to Cache: %v \n", key)
+
+}
+
+// getCachedBindings retrieves all bindings from the cache
+func (c *resourceCache) getCachedBindings() map[string]*facade.Binding {
+	return c.bindings
+}
+
+// getBindingFromCache retrieves a specific binding from the cache
+func (c *resourceCache) getBindingFromCache(key string) (*facade.Binding, bool) {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	binding, found := c.bindings[key]
+	// TODO :remove After internal review
+	fmt.Printf("Got the binding from Cache: %v \n", binding)
+	return binding, found
+}
+
+// updateBindingInCache updates a specific binding in the cache
+func (c *resourceCache) updateBindingInCache(owner string, parameters map[string]interface{}, generation int64) (status bool) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	//update if the instance is found in the cache
+	//update all the struct variables if they are not nil or empty
+	binding, found := c.bindings[owner]
+	if found {
+		if parameters != nil {
+			binding.ParameterHash = facade.ObjectHash(parameters)
+		}
+		if generation != 0 {
+			binding.Generation = generation
+		}
+		c.bindings[owner] = binding
+		return true
+
+	}
+	return false
+}
+
+// -----------------------------------------------------------------------------------------------
+// Instances
+// -----------------------------------------------------------------------------------------------
+
+// addInstanceInCache stores an instance to the cache
+func (c *resourceCache) addInstanceInCache(key string, instance *facade.Instance) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	// TODO :remove After internal review
+	fmt.Printf("Added the instance to Cache: %v \n", instance)
+	c.instances[key] = instance
+}
+
+// deleteInstanceFromCache deletes a specific instance from the cache
+func (c *resourceCache) deleteInstanceFromCache(key string) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	delete(c.instances, key)
+	// TODO :remove After internal review
+	fmt.Printf("deleted the instance from Cache: %v \n", key)
+
+}
+
+// getCachedInstances retrieves all instances from the cache
+func (c *resourceCache) getCachedInstances() map[string]*facade.Instance {
+	return c.instances
+}
```

Comment:
I think that if another reconciliation tries to modify the cache by adding or deleting instances we can have a race condition. I would add an RLock() so we are sure that it is safe to access when retrieving instances.

I think the same about the spaces and bindings.

Reply:
i deleted the unused functions